## Iron Drives Anabolic Metabolism Through Active Histone Demethylation and mTORC1
 
Jason S. Shapiro#, Hsiang-Chun Chang#, Yuki Tatekoshi, Zibo Zhao, Zohra Sattar Waxali, Bong Jin Hong, Haimei Chen, Justin A. Geier, Elizabeth T. Bartom, Adam De Jesus, Farnaz K. Nejad, Amir Mahmoodzadeh, Tatsuya Sato, Lucia Ramos-Alonso, Antonia Maria Romero, Maria Teresa Martinez-Pastor, Shang-Chuan Jiang, Shiv K. Sah-Teli, Liming Li, David Bentrem, Gary Lopaschuk, Issam Ben-Sahra, Thomas V. Oâ€™Halloran, Ali Shilatifard, Sergi Puig, Joy Bergelson, Peppi Koivunen and Hossein Ardehali*

#These authors contributed equally to this work

*Corresponding author. Email: h-ardehali@northwestern.edu


### Abstract
All eukaryotic cells require a minimal iron threshold to sustain anabolic metabolism. However, the mechanisms by which cells sense iron to regulate anabolic processes are unclear. Here, we report a previously undescribed eukaryotic pathway for iron sensing in which molecular iron is required to sustain active histone demethylation and maintain the expression of critical components of the pro-anabolic mTORC1 pathway. Specifically, we identify the iron-binding histone-demethylase KDM3B as an intrinsic iron sensor that regulates mTORC1 activity by demethylating H3K9me2 at enhancers of a high-affinity leucine transporter, LAT3, and RAPTOR. By directly suppressing leucine availability and RAPTOR levels, iron deficiency (ID) supersedes other nutrient inputs into mTORC1. This process occurs in vivo and is not an indirect effect by canonical iron-utilizing pathways. Because ancestral eukaryotes share homologues of KDMs and mTORC1 core components, this pathway likely predated the emergence of the other kingdom-specific nutrient sensors for mTORC1.

The source code linked to this file were used for analyzing the NGS data from the above study. All of the code needed to reproduce the published results, including the source code for the Ceto pipeline and its associated scripts are available at the Github sites: [https://github.com/ebartom/NGSbartom](https://github.com/ebartom/NGSbartom) and [https://github.com/ebartom/Shapiroetal2023](https://github.com/ebartom/Shapiroetal2023).

NGS data generated for this study are available at the Gene Expression Omnibus (GEO) under accession number [GSE214019](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE214019). Instructions for accessing additional files needed to run the code are annotated within source code.

Please contact the corresponding author and Dr. Elizabeth T. Bartom (email: ebartom@northwestern.edu) for any inquiries pertaining to this source code.


